src/main/java/org/folio/processing/mapping/mapper/reader/record/MarcRecordReader.java

```diff
@@ -1,61 +1,180 @@
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
 import org.folio.DataImportEventPayload;
 import org.folio.processing.mapping.mapper.reader.Reader;
+import org.folio.processing.value.BooleanValue;
+import org.folio.processing.value.ListValue;
 import org.folio.processing.value.MissingValue;
+import org.folio.processing.value.RepeatableFieldValue;
 import org.folio.processing.value.StringValue;
 import org.folio.processing.value.Value;
 import org.folio.rest.jaxrs.model.EntityType;
+import org.folio.rest.jaxrs.model.MappingRule;
+import org.folio.rest.jaxrs.model.RepeatableSubfieldMapping;
+import org.marc4j.MarcJsonReader;
+import org.marc4j.MarcReader;
+import org.marc4j.marc.Record;
+import org.marc4j.marc.VariableField;
+import org.marc4j.marc.impl.ControlFieldImpl;
+import org.marc4j.marc.impl.DataFieldImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import java.io.IOException;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static org.apache.commons.lang3.StringUtils.EMPTY;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
 
 @SuppressWarnings("all")
 public class MarcRecordReader implements Reader {
-  private static final String MARC_FIELDS_POINTER = "/fields";
+  private final static Pattern MARC_PATTERN = Pattern.compile("(^[0-9]{3}(\\$[a-z]$){0,2})");
+  private final static Pattern STRING_VALUE_PATTERN = Pattern.compile("(\"[^\"]+\")");
+  private final static String WHITESPACE_DIVIDER = "\\s(?=(?:[^'\"`]*(['\"`])[^'\"`]*\\1)*[^'\"`]*$)";
+  private final static String EXPRESSIONS_DIVIDER = "; else ";
+  private final static String EXPRESSIONS_ARRAY = "[]";
+  private final static String EXPRESSIONS_QUOTE = "\"";
+  private static final Logger LOGGER = LoggerFactory.getLogger(MarcRecordReader.class);
   private EntityType entityType;
-  private JsonNode fieldsNode = null;
+  private Record marcRecord;
 
   MarcRecordReader(EntityType entityType) {
     this.entityType = entityType;
   }
 
   @Override
-  public void initialize(DataImportEventPayload eventPayload) throws IOException {
-    if (eventPayload.getContext().containsKey(entityType.value())) {
-      String stringRecord = eventPayload.getContext().get(entityType.value());
-      this.fieldsNode = new ObjectMapper().readTree(stringRecord).at(MARC_FIELDS_POINTER);
-    } else {
-      throw new IllegalArgumentException("Can not initialize MarcRecordReader, no suitable entity type found in event payload");
+  public void initialize(DataImportEventPayload eventPayload) throws RuntimeException {
+    try {
+      if (eventPayload.getContext().containsKey(entityType.value())) {
```

Comment:
eventPayload.getContext != null ? 
Looks like HashMap types are not initialized by default in generated schemas